### PR TITLE
Support multiple EXTNAME in compressed images

### PR DESCRIPTION
**Pull request by timj**
### Description

I have some old files that set `EXTNAME` header to both `COMPRESSED_IMAGE` and an explicit value. Previously the code would use the first value for the name and then delete all values from the header returned to the user (because it contained `COMPRESSED_IMAGE`). This makes it impossible for the user to discover the `EXTNAME` of interest.

This pull request is to address this by checking for multiple EXTNAME and removing only the default.
It leaves the other behind and uses it for `.name` property.

### CODE REVIEW

File: astropy/io/fits/hdu/compressed.py [hunk 1]
```diff
@@ -666,6 +668,24 @@ def __init__(self, data=None, header=None, name=None,
...
+    @property
+    def name(self):
+        if 'EXTNAME' not in self._header:
+            return self._default_name
+
+        name = self._default_name
+
+        # If there are multiple EXTNAME values use the first that
+        # is not COMPRESSED_IMAGE
+        indices = self._header._keyword_indices['EXTNAME']
```
Comment (timj) [inline astropy/io/fits/hdu/compressed.py:680]: I'm not entirely sure if I'm allowed to use the `_keyword_indices` method but I couldn't find a public API that would let me find which cards to test.

**Comment by timj** [comment]
I need to think about adding tests but I wanted something submitted first in case the whole approach is thrown out.

This PR came out of a Slack discussion with @saimn.

**Comment by timj** [comment]
I do see the test failure. Seems like the `name` setter is not inherited. Will take a look later.

**Comment by timj** [comment]
I fixed the problem with the setter but I don't understand why the python3.8 test is failing with samp.

**Comment by pllim** [comment]
@timj , the `samp` failure has been spotted in other PRs too and is unrelated to your changes. Fortunately(?), it is transient, so I restarted the CI here, hoping that it would go away. 🤞

File: astropy/io/fits/hdu/compressed.py [hunk 2]
```diff
@@ -694,6 +718,31 @@ def match_header(cls, header):
...
+                image_header['EXTNAME'] == 'COMPRESSED_IMAGE'):
+            # In some rare cases there might be a COMPRESSED_IMAGE in
+            # the header and non-default EXTNAME. Only delete the default.
+            indices = image_header._keyword_indices['EXTNAME']
+            to_remove = []
+            for index in indices:
+                if image_header[index] == 'COMPRESSED_IMAGE':
+                    to_remove.append(index)
+            for index in to_remove:
+                del image_header[index]
```
Comment (saimn) [inline astropy/io/fits/hdu/compressed.py:744]: This doesn't work if COMPRESSED_IMAGE is present more than once. Because when you remove the first occurrence it modifies the indices of indices, so the indices that you have in `to_remove` are no more valid.
The solution is probably to delete just the first COMPRESSED_IMAGE found, if there are more than one, well ...

File: astropy/io/fits/tests/test_image.py [hunk 3]
```diff
@@ -1732,6 +1732,13 @@ def test_compressed_header_missing_znaxis(self):
         with pytest.raises(KeyError):
             comp_hdu.compressed_data
 
+    def test_compressed_header_double_extname(self):
+        """Test that a double EXTNAME with one default value does not
+        mask the non-default value."""
+        with fits.open(self.data('double_ext.fits')) as hdul:
+            assert hdul[1].name == "ccd00"
+            assert "EXTNAME" in hdul[1].header
```
Comment (saimn) [inline astropy/io/fits/tests/test_image.py:1740]: Could you also test that if you write the file the COMPRESSED_IMAGE card is removed ?

File: astropy/io/fits/hdu/compressed.py (hunk 2 above)
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:744]: I meant to say:
```python
for index in reversed(to_remove):
   del image_header[index]
```
since presumably that does work without causing any reindexing problems.

Comment (timj) [reply astropy/io/fits/hdu/compressed.py:744]: if somehow we have two COMPRESSED_IMAGE entries and then the real one we'd still want to delete both and if we didn't it would report COMPRESSED_IMAGE as now. I don't think what I've got is going to cause any particular slowdown because in almost every case it would only loop once and even if we had two COMPRESSED_IMAGE and a real one it would only loop three times.

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (timj) [reply astropy/io/fits/tests/test_image.py:1740]: I presumably can't use astropy to read the header though because the code here is going to remove the COMPRESSED_IMAGE card on read even if it was written out.

File: astropy/io/fits/hdu/compressed.py (hunk 2 above)
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:746]: In fact the test file I have included here has 4 EXTNAME headers in it (two of them are COMPRESSED_IMAGE).

File: astropy/io/fits/hdu/compressed.py [hunk 4]
```diff
@@ -666,6 +668,28 @@ def __init__(self, data=None, header=None, name=None,
...
+        name = self._default_name
+
+        # If there are multiple EXTNAME values use the first that
+        # is not COMPRESSED_IMAGE
+        indices = self._header._keyword_indices['EXTNAME']
+        for index in indices:
+            if self._header[index] != 'COMPRESSED_IMAGE':
+                # Convert the value to a string to be flexible in some pathological
+                # cases (see ticket #96)
+                name = str(self._header[index])
```
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:685]: Yes, I think returning the first non-COMPRESSED_IMAGE is the right thing here.

File: astropy/io/fits/hdu/compressed.py [hunk 5]
```diff
@@ -666,6 +668,28 @@ def __init__(self, data=None, header=None, name=None,
         self._orig_bscale = self._bscale
         self._orig_bitpix = self._bitpix
 
+    @property
+    def name(self):
```
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:672]: Looking at it I think we might need a special setter as well as getter. The issue is that with the current implementation on main you can do (on a compressed header):

```python
>>> hdu.name = "FOO"
>>> hdu.header["EXTNAME"] = "BAR"
```
and when you do `hdu.writeto("new.fits")` you end up with "FOO" in the header and "BAR" is dropped.  I think this is because the name setter sets `._header` but the `hdu.header` is `._image_header`. I think this needs to be changed to make `.name` use `_image_header`.

File: astropy/io/fits/hdu/compressed.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:749]: 4 EXTNAME :scream: 
I guess the indices in `_keyword_indices` should be ordered, at least that's the case when you read a file, though you could still manually insert a duplicate card before another one. Maybe a `reversed(sorted(to_remove))` to be sure ?

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/tests/test_image.py:1734]: I think the cards would still be in `_header` ?

**Review by saimn — COMMENTED** [review]
Hey @timj,
Sorry I forgot about this one. Do you think you could finalize before end of next week (feature freeze for 4.3) ?

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/tests/test_image.py:1734]: Also, maybe you could add a few more EXTNAME/COMPRESSED_IMAGE to your test file ? (I haven't checked, maybe that's already the case)

**Comment by timj** [comment]
@saimn sorry. I'll see if I can have another go at it. I had thrown up my hands a bit when I realized how confused I was with the header and image_header stuff and how it was interacting with the base class and was starting to despair. It seemed like every time I tried to clean something up I made it worse.

File: astropy/io/fits/hdu/compressed.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py]: I changed it to use `self.header` in both places. That should cause it to sync to both `self._image_header` and `self._header` and so be consistent.

File: astropy/io/fits/hdu/compressed.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py]: I think this is now moot because `name` reads `self.header` which has already been fixed up.

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (timj) [reply astropy/io/fits/tests/test_image.py:1734]: With the new code I actually strip them from `_header` on read so I can't tell either way. I'm not entirely clear whether `_header` is meant to be treated as read-only since the `CompImageHeader` is clearly designed to write to `_header` for any key that is not one of the special ones.

**Comment by timj** [comment]
I'm not entirely sure what I've done that is breaking the checksum tests. I'm having trouble running the tests locally at the moment (tox is filling up my tmp space) but I'll try again tomorrow.

**Comment by timj** [comment]
I think the checksum tests are failing because `hdu.name = x` forces updates of the synced header (table header and image header) *and* the code for writing the table header was calling `.name`. It didn't really need to do that because the code was forcing EXTNAME into `_header` and then forcing `.name` to sync that value back to `_header`. I was wondering about the purpose of that and removing the extra setting fixes the problem.

**Comment by timj** [comment]
@saimn I finally have tests passing so I think it's ready for another look. The CI failure is in codecov where I have seemingly made test coverage worse (even though I expanded the tests).

The `name` setter is now copied from the base class rather than moving the shared code to a separate method. I'm not sure which way you'd rather I went but it's an easy fix either way.

**Comment by timj** [comment]
I updated the tests to improve code coverage and all tests pass.

From my point of view the remaining item for discussion is whether I should refactor the base class `name` setter to allow reuse in the subclass? At present one is using `.header` and the other is using `._header` so the only way to encourage reuse is to have them both call a static method that passes in the header to be updated.

I think it's right for the table header creation code to not force a default COMPRESSED_IMAGE EXTNAME into the image header but it's not clear if we should be forcing the non-default name specified by the user into the image header as well since on read that would propagate through. The current PR preserves the status quo of not changing the image header at all.

**Comment by saimn** [comment]
Thanks @timj, I will try to review soon.

> whether I should refactor the base class name setter to allow reuse in the subclass? At present one is using .header and the other is using ._header so the only way to encourage reuse is to have them both call a static method that passes in the header to be updated.

Yes I'm not sure, `hdu.base` is using `._header` everywhere so replacing it only in the name getter/setter would be confusing. I need to have a look at the code first to get a better idea.

**Comment by timj** [comment]
Thanks. Most of the pain here is because of the table vs image header. I'm starting to wonder if it would have been better if the subclass had used `._header` as the synced image header and added a new `._table_header` for the behind-the-scenes "what's written to the file" header. It's entirely possible I've been focused so much on EXTNAME that I've missed some other complication. The previous code led to anomalous situations where you could end up with`hdu.name` being inconsistent with  `hdu.header["EXTNAME"]`.

File: astropy/io/fits/hdu/compressed.py [hunk 6]
```diff
@@ -672,6 +675,57 @@ def __init__(self, data=None, header=None, name=None,
...
+        """
+        if 'EXTNAME' in header:
+            indices = header._keyword_indices['EXTNAME']
+            # Only continue if there is more than one found
+            n_extname = len(indices)
+            if n_extname > 1:
+                extnames_to_remove = []
+                for index in indices:
+                    if header[index] == self._default_name:
+                        extnames_to_remove.append(index)
```
Comment (saimn) [inline astropy/io/fits/hdu/compressed.py:694]: Maybe with a list comprehension ?
```
extnames_to_remove = [index for index in indices
                      if header[index] == self._default_name]
```

File: astropy/io/fits/hdu/compressed.py [hunk 7]
```diff
@@ -672,6 +675,57 @@ def __init__(self, data=None, header=None, name=None,
...
+        if not conf.extension_name_case_sensitive:
+            value = value.upper()
+        if 'EXTNAME' in self.header:
+            # Expectation is that all EXTNAME headers found will be
+            # replaced in both _header and _image_header.
+            # Up to this point _header might be the original header
+            # but now we can call the code to remove unnecessary
+            # COMPRESSED_IMAGE EXTNAME since we know we are setting a
+            # new value.
+            self._remove_unnecessary_default_extnames(self._header)
```
Comment (saimn) [inline astropy/io/fits/hdu/compressed.py:724]: I think this could be removed (see below).

File: astropy/io/fits/hdu/compressed.py [hunk 8]
```diff
@@ -1559,10 +1613,14 @@ def header(self):
             image_header.set('DATASUM', self._header['ZDATASUM'],
                              self._header.comments['ZDATASUM'])
 
+        # Clean up any possible doubled EXTNAME keywords that use
+        # the default.
+        self._remove_unnecessary_default_extnames(image_header)
```
Comment (saimn) [inline astropy/io/fits/hdu/compressed.py:1618]: I think we could do that before copying `_header` (`image_header = self._header.copy()`), this way both `._header` and `_image_header` would be cleaned and calling `_remove_unnecessary_default_extnames` in the name setter would no more be necessary. I tried locally and it seems to work.

**Comment by saimn** [comment]
Re table vs image header: Yes you're right, and this code is really hard to follow. The more general issue is that `CompImageHDU` is a subclass of `BinTableHDU`. It would probably be easier to subclass `ImageHDU` and handle internally a `BinTableHDU` but that's a significant work (#9238).

File: astropy/io/fits/hdu/compressed.py [hunk 9]
```diff
@@ -791,15 +848,12 @@ def _update_header_data(self, image_header,
...
-            name = 'COMPRESSED_IMAGE'
+            name = self._default_name
 
         if name:
             self._header.set('EXTNAME', name,
                              'name of this binary table extension',
                              after='TFIELDS')
-            self.name = name
-        else:
-            self.name = self._header['EXTNAME']
```
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:802]: `self.name` now sets `.header` which sets `._image_header` and `._header`. The assignment here was always suspect because it copies a EXTNAME value from `._header` and then ended up setting it back in `._header`.  It seemed always wrong. I can leave it in but that would lead to the _image_header getting the EXTNAME and that changes the checksum calculation (tests fail). Maybe that's the right thing to do but we should not be doing it if we are deciding to write COMPRESSED_IMAGE EXTNAME in (the default determined a few lines earlier). So I'm fine with retaining `self.name` and changing checksums in image_header so long as we don't do it for COMPRESSED_IMAGE.

File: astropy/io/fits/hdu/compressed.py (hunk 8 above)
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:1618]: I did originally do that but then I saw some comments about `_header` sometimes being read-only (depending on how the header is mapped from a file) and decided that it was safer to defer updating `_header` until the user actively asks for the header to be changed.

File: astropy/io/fits/hdu/compressed.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:802]: But, `_image_header` should also get the EXTNAME, otherwise it would not appear with `.header` ?

Not sure if that's related with those lines but I'm seeing some weird behavior:
```
In [5]: hdu = fits.CompImageHDU(data=np.arange(10), name='FOO')

In [6]: hdu._header['EXTNAME']
Out[6]: 'FOO'

In [7]: hdu.header  ## EXTNAME doesn't appear
Out[7]: 
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                   64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                   10                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               

In [8]: hdu.name = 'BAR'

In [9]: hdu.header  ## OK
Out[9]: 
XTENSION= 'IMAGE   '           / Image extension                                
BITPIX  =                   64 / array data type                                
NAXIS   =                    1 / number of array dimensions                     
NAXIS1  =                   10                                                  
PCOUNT  =                    0 / number of parameters                           
GCOUNT  =                    1 / number of groups                               
EXTNAME = 'BAR     '           / extension name                                 

In [10]: hdu._header  ## Duplicate EXTNAME
Out[10]: 
...
ZVAL2   =                    4 / bytes per pixel (1, 2, 4, or 8)                
EXTNAME = 'BAR     '           / extension name                                 
EXTNAME = 'BAR     '           / extension name                                 
```

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:802]: The same happen with the default EXTNAME (`fits.CompImageHDU(data=np.arange(10))`), so I guess something to do with `CompImageHeader`.

[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:802]: As a further comment the code before was doing:

if name is defined:
    set EXTNAME in _header to "name"
    set EXTNAME in _header via .name to "name"
else:
    set EXTNAME in _header via .name to "name"

The call of .name was completely irrelevant and image_header was not involved. Currently that is also true but because the image_header is not having EXTNAME set the `hdu.name` call will not return anything and will return the default of COMPRESSED_IMAGE. Previously hdu.name only looked at _header but now it looks at `.header`.  Option 1 is to do what I say above and update the `.header` and update the checksums. The other option is to allow `hdu.name` to first look in `.header` or else fallback to `._header` if they are out of sync.

File: astropy/io/fits/hdu/compressed.py (hunk 8 above)
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:1618]: Hmm, I don't think a `Header` could be read-only, but maybe I missed something.

File: astropy/io/fits/hdu/compressed.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:802]: Yeah OK, well this is a mess :grin:, sorry for the trap. 
I didn't follow about the checksum issue, but if I understand correctly the checksum should change if EXTNAME is changed so this is a good thing isn't it ? (fixing another bug)

[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:802]: Yes, so previously only `._header` was changing and not `.header` so yes, I was wondering if it was an actual bug that I fixed causing the checksums to change. I'll restore setting `.header` and fix the checksums but I'm fairly certain I don't want the COMPRESSED_IMAGE value to leak into `.header` since we spend all our time trying to make it disappear.

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/compressed.py:802]: I agree about COMPRESSED_IMAGE, it should only be added when writing a file if EXTNAME is not set.

File: astropy/io/fits/hdu/compressed.py (hunk 8 above)
Comment (timj) [reply astropy/io/fits/hdu/compressed.py]: I've moved it to before the copy.

File: astropy/io/fits/hdu/compressed.py (hunk 7 above)
Comment (timj) [reply astropy/io/fits/hdu/compressed.py]: Removed.

File: astropy/io/fits/hdu/compressed.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (timj) [reply astropy/io/fits/hdu/compressed.py:802]: I have now done this. The checksums don't have to be fixed because in this specific example it was using the COMPRESSED_IMAGE default and I do not update .header in that situation.

**Comment by saimn** [comment]
Thanks for the hard work on this @timj :), it certainly improves things and looks good to me.

**Review by saimn — APPROVED** [review]

**Comment by timj** [comment]
Thanks it was much more convoluted than I ever expected when I started.

**Comment by pllim** [comment]
> much more convoluted than I ever expected when I started

Welcome to FITS. 😹

**Comment by embray** [comment]
I'm not sure I understand or agree with how this was handled.  I don't think any standard dictates that a compressed image HDU must have an `EXTNAME` of `COMPRESSED_IMAGE` by default.  It appears to be a convention carried over from CFITSIO.

More importantly, the FITS standard does not have any provision for "multiple EXTNAMEs", and the fact that there were ever multiple EXTNAMEs in a compressed image header is a bug to begin with is it not?

For reference CFITSIO (and `imcopy`) replace this default EXTNAME if you compress an HDU that already has an EXTNAME.  It doesn't create a duplicate.

I guess, if there are FITS files in the wild created this way due to a buggy Astropy/PyFITS, then we should account for that case... :(

But I wonder if the original bug has even been fixed.

**Comment by timj** [comment]
The reason I had to deal with this was that there is data in the CFHT archive that has this problem and reading it was stripping the real EXTNAME value. Other systems were working because the second EXTNAME in the file effectively overwrote the first one but in astropy the first value is picked preferentially. Yes the compressed image header class is a maze of twisty passages.

**Comment by embray** [comment]
I believe that said data must have been written by Astropy, since I found a bug that would cause exactly that: Duplicate EXTNAME values.  I guess if Astropy caused the bug we have to eat our own garbage and handle it, so thank you for the fix.  Next step is to fix the original issue (I am working on a fix but it's a bit tricky...)
